### PR TITLE
Ding/impl mark region in source 1

### PR DESCRIPTION
**Pull request by dingyuchen**

### CODE REVIEW

File: src/virtual-machines/source-1-with-mark-region-gc.js [hunk 1]
```diff
@@ -0,0 +1,1888 @@
...
+  J = A;
+  K = B;
+  if (BUMP_HEAD + K > BUMP_TAIL) {
+    display("Sweeeeep!");
+    // mark and free
+    MARK();
+    FREE_REGION();
+    // get new bump and limit pointers
+    // ALLOCATE_BUMP_HEAD();
+    // ALLOCATE_BUMP_TAIL();
```
Comment (moziliar) [inline src/virtual-machines/source-1-with-mark-region-gc.js]: Only mark and free region when there's no more free blocks. Do not do it prematurely.

File: src/virtual-machines/source-1-with-mark-region-gc.js [hunk 2]
```diff
@@ -0,0 +1,1888 @@
...
+  }
+}
+
+// expects object address in A
+// returns line node address in RES
+function GET_LINE() {
+  // block address in C
+  C = math_floor(A / BLOCK_SIZE) * BLOCK_SIZE;
+  // abuse RES to store start of addressable space
+  RES = HEAP[HEAP[C + FIRST_CHILD_SLOT] + LINE_ADDRESS_SLOT];
```
Comment (moziliar) [inline src/virtual-machines/source-1-with-mark-region-gc.js]: Just use some register plz... let's keep RES solely for result of a VM instruction

File: src/virtual-machines/source-1-with-mark-region-gc.js [hunk 3]
```diff
@@ -0,0 +1,1888 @@
...
+    } else {
+    }
+  }
+}
+
+// expects object address in A
+// returns line node address in RES
+function GET_LINE() {
+  // block address in C
+  C = math_floor(A / BLOCK_SIZE) * BLOCK_SIZE;
```
Comment (moziliar) [inline src/virtual-machines/source-1-with-mark-region-gc.js]: You could keep track of current block with a header register like CURR_BLOCK or something?

File: src/virtual-machines/source-1-with-mark-region-gc.js (hunk 2 above)
Comment (moziliar) [inline src/virtual-machines/source-1-with-mark-region-gc.js]:
```suggestion
  RES = HEAP[C + HEAP[C + FIRST_CHILD_SLOT] + LINE_ADDRESS_SLOT];
```
Is there a missing C?

File: src/virtual-machines/source-1-with-mark-region-gc.js [hunk 4]
```diff
@@ -0,0 +1,1888 @@
...
+  // abuse RES to store start of addressable space
+  RES = HEAP[HEAP[C + FIRST_CHILD_SLOT] + LINE_ADDRESS_SLOT];
+  // line number in A
+  A = math_floor((A - RES) / LINE_SIZE);
```
Comment (moziliar) [inline src/virtual-machines/source-1-with-mark-region-gc.js]: Likewise, could we do this with a CURR_LINE register?

**Review by moziliar — CHANGES_REQUESTED** [review]

File: src/virtual-machines/source-1-with-mark-region-gc.js (hunk 3 above)
Comment (dingyuchen) [reply src/virtual-machines/source-1-with-mark-region-gc.js]: `GET_LINE()` is agnostic and may not always correspond with `BUMP` pointers. We can use it to always follow the `SCAN` pointer, then we will need to nail down design decision first.

File: src/virtual-machines/source-1-with-mark-region-gc.js (hunk 1 above)
Comment (dingyuchen) [reply src/virtual-machines/source-1-with-mark-region-gc.js]: Will add this in for next PR.

File: src/virtual-machines/source-1-with-mark-region-gc.js (hunk 4 above)
Comment (dingyuchen) [reply src/virtual-machines/source-1-with-mark-region-gc.js]: Similar argument for `CURR_BLOCK`

File: src/virtual-machines/source-1-with-mark-region-gc.js (hunk 3 above)
Comment (moziliar) [reply src/virtual-machines/source-1-with-mark-region-gc.js]: Ok point taken

**Review by moziliar — APPROVED** [review]
LGTM

**Pull request merged**